Feature/add node stores

seawatts opened the pull request:
Added node store
Added unit tests for node store
TODO: TenantID is not implemented
TODO: Have not refactored node code out of cluster store

Commit:
Added: Node store service
Added: Node store views
Added: Tests for get node
Added: Tests for delete node
Added: Function to write multiple nodes.
Added: Node exists in node store
Added: Node set and insert to node store
Added: Tests for deleting node as user and admin
Added: Tests for testing update node in store

seawatts commented:
Actually hold on a sec, broke some tests.

wolf31o2 commented on provisioner/Gemfile:
```diff
@@ -8,6 +8,10 @@ group :dependencies do
+  gem 'knife-joyent', '= 0.4.8'
+  gem 'knife-openstack', '= 0.8.1'
+  gem 'knife-rackspace', '= 0.9.1'
+
```
I don't really like this, since it is technically incorrect, but let's go ahead and up them in until we clean things up a bit more and rearrange the plugins, since they won't be shipping with the provisioner in 0.9.8 and beyond.

seawatts commented:
Reverted a ton of files that were not needed to try to keep the PR smaller.
Also fixed some formatting issues that were unnecessary.
All tests passing now. 

Ready to be merged

albertshau commented on server/src/main/java/com/continuuity/loom/store/node/NodeStoreView.java:
```diff
@@ -0,0 +1,63 @@
+   * @return Set of all nodes belonging to specified cluster. Empty if no cluster or cluster nodes exist.
+   * @throws java.io.IOException if there was a problem getting the cluster nodes.
+   */
+  Set<Node> getClusterNodes(String clusterId) throws IOException;
```
should this be in the node store?  Seems like it fits better in the cluster store plus all the implementations here just return null.

albertshau commented on server/src/main/java/com/continuuity/loom/store/node/NodeStore.java:
```diff
@@ -0,0 +1,7 @@
+/**
+ * A full view of the node store for management purposes.
+ */
+public interface NodeStore extends NodeStoreView {}
```
we could remove this interface and just have the view since right now its just another layer of indirection. Unless there are plans in the near future to add methods here?

albertshau commented on server/src/main/java/com/continuuity/loom/store/node/SQLNodeStoreService.java:
```diff
@@ -0,0 +1,80 @@
+
+  @Override
+  protected void startUp() throws Exception {
+    boolean created = DBHelper.createDerbyTableIfNotExists("CREATE TABLE nodes ( " +
```
should be wrapped in a check that we're using the embedded derby db (method in dbConnectionPool)

albertshau commented on server/src/main/java/com/continuuity/loom/store/node/SQLUserNodeStoreView.java:
```diff
@@ -0,0 +1,72 @@
+package com.continuuity.loom.store.node;
```
missing copyright

albertshau replied:
same goes for other new files

albertshau commented on server/src/test/java/com/continuuity/loom/store/node/SQLNodeStoreTest.java:
```diff
@@ -0,0 +1,59 @@
+
+  @BeforeClass
+  public static void setupSQLNodeStoreTest() throws SQLException {
+    Configuration sqlConf = Configuration.create(); sqlConf.set(Constants.JDBC_DRIVER, "org.apache.derby.jdbc.EmbeddedDriver"); sqlConf.set(Constants.JDBC_CONNECTION_STRING, "jdbc:derby:memory:loom;create=true");
```
should have 1 statement per line

seawatts replied:
Ah, weird, IntelliJ must have done that on a reformat. =/

albertshau commented on server/src/test/java/com/continuuity/loom/store/node/NodeStoreTest.java:
```diff
@@ -0,0 +1,165 @@
+  //  private static final Account tenant1_user2 = new Account("user2", "tenant1");
+  private static final Account tenant1_admin = new Account(Constants.ADMIN_USER, "tenant1");
+  //  private static final Account tenant2_user1 = new Account("user1", "tenant2");
+  //  private static final Account tenant2_admin = new Account(Constants.ADMIN_USER, "tenant2");
```
if we don't need them, we should delete instead of commenting

albertshau commented on server/src/main/java/com/continuuity/loom/store/node/SQLAdminNodeStoreView.java:
```diff
@@ -0,0 +1,72 @@
+ * The node store as viewed by a tenant admin. A tenant admin can read, write, and delete any cluster
+ * within the tenant.
+ */
+public class SQLAdminNodeStoreView extends BaseSQLNodeStoreView {
```
I'm assuming the plan is to first refactor nodes into their own classes and then add user and tenant to the database table?

seawatts replied:
Yes, I was planning on doing that in a future branch. Trying to keep the PR's smallish =)

albertshau commented on server/src/test/java/com/continuuity/loom/store/node/NodeStoreTest.java:
```diff
@@ -0,0 +1,165 @@
+    view.writeNodes(mockNodes);
+    assertNodeSetWritten(mockNodes);
+    Set<Node> actualNodes = view.getAllNodes();
+    assertNodeSetAgainstNodeSet(actualNodes, mockNodes);
```
I think you should be able to compare the sets directly instead of converting to arrays.

albertshau commented:
few small comments, looking good to merge after those.

Also, we will need a CLA from the company before merging, will try and get one to you guys later today.

Commit:
Updated: Copyright
Fixed: Code style issues
Deleted: Unused commented code
Added: DBConnection check for embedded DB
Removed: getClusterNodes from NodeStoreView. Should be in ClusterStoreView

seawatts commented:
Alright, fixed up some code review comments.

seawatts commented:
Can someone help me figure out why this is not creating the loom schema when running the NodeStoreTest

albertshau commented:
probably it's because the node service isn't being started. Try calling startAndWait() when setting up the test.

albertshau commented:
by the way, looking into other failures in travis at the moment so you may see some other failures

Commit:
Fixed: Failing test by starting node service before running
Removed: Redundant createStatement call in SQLNodeStoreService

albertshau commented:
looks good, just need the signed CLA and we can merge

nitinmotgi commented:
:+1: Approved to merge.
